### ISSUE TITLE
extends null

### ISSUE THREAD

**Issue by zloirock** [tc39/ecma262#543]
It does not allow me to sleep for a year. It was broken after the last class reform. [See the last sentence.](https://github.com/tc39/ecma262/blob/703238399d339ceb7f894f4733626f601d42d05e/workingdocs/ES6-super-construct%3Dproposal.md) For usage `this` should be called `super`, but we can't call `null`. [This issue was closed.](https://github.com/tc39/ecma262/issues/22) Now added many serious changes for ES6, so maybe makes sense fix `extends null` like it was proposed in #22? Usage `this` in extended from `null` constructors without `super` or a special case for `super` in those constructors should not break any current code, but should make `extends null` useful instead of current useless garbage in the spec.

**Comment by bterlson** [comment]
@zloirock not sleeping is definitely a bug. We should fix this. I think @allenwb's suggestion in #22 is sound (modulo early error restrictions, but I'm not seeing any).

**Comment by zloirock** [comment]
@bterlson special case for usage `super` in those constructors should remove early error and default constructor issues.

**Comment by allenwb** [comment]
I agree.  I think it is a specification bug that `extend null` classes were marked as "derived"

**Comment by loganfsmyth** [comment]
@bterlson Assuming I'm following this right, this still means the existing spec behavior of

```
class Foo extends null {
    constructor(){
        super();
    }
}
```

will throw because `super()` can't be used in an `extends null` class, correct? If so, it seems like step 10 should also be updated, since this will now initialize `this` by default, but still default to calling `super()` when no constructor is given.

I'll admit, auto-initializing `this` seems unfortunate to me though, it seems much more consistent to always require `super()` to initialize `this` even in cases where it is `null`, since "use `super()` when you have `extends`" is a more straightforward thing to understand from an end-user stand-point.

**Comment by bterlson** [comment]
Yep, good catch, overlooked it (and fixed via a35f0f3). I don't think super makes sense here since there is no super constructor.

**Comment by littledan** [comment]
The change here makes sense to me, but I missed this thread until the test262 test went in and V8 failed it. I'm wondering, for normative changes like this, would it make sense to summarize them at TC39 (or some other place which really highlights it), and get some implementation feedback (or documentation that an implementation already does this) before committing?
